## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 516705`
Commit: [`991cd371aeb3acd8675ea9835010ebc9e7f22fc4`](https://github.com/NixOS/nixpkgs/commit/991cd371aeb3acd8675ea9835010ebc9e7f22fc4) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/991cd371aeb3acd8675ea9835010ebc9e7f22fc4..pull/516705/head))
Merge: [`9afe61059ed12aaaee3529e5dd51e9cdaad3e879`](https://github.com/NixOS/nixpkgs/commit/9afe61059ed12aaaee3529e5dd51e9cdaad3e879)

Logs: https://github.com/kilyanni/nixpkgs-review-gha/actions/runs/25353578877


---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>wasmer</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>wasmer</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>wasmer</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>wasmer</summary>
<pre>LLVM version: .


--------------
RULE EXECUTION
--------------


install -Dm755 target/release/wasmer /nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/bin/wasmer
for header in lib/c-api/*.h; do install -Dm644 &quot;$header&quot; /nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/include/$(basename $header); done
install -Dm644 lib/c-api/README.md /nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/include/wasmer-README.md
pkgver=$(cargo pkgid --manifest-path lib/c-api/Cargo.toml | sed &#x27;s/^.*wasmer-c-api@//&#x27;) &amp;&amp; \
shortver=&quot;${pkgver%.*}&quot; &amp;&amp; \
majorver=&quot;${shortver%.*}&quot; &amp;&amp; \
install -Dm755 target/release/libwasmer.so &quot;/nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/lib/libwasmer.so.$pkgver&quot; &amp;&amp; \
ln -sf &quot;libwasmer.so.$pkgver&quot; &quot;/nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/lib/libwasmer.so.$shortver&quot; &amp;&amp; \
ln -sf &quot;libwasmer.so.$pkgver&quot; &quot;/nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/lib/libwasmer.so.$majorver&quot; &amp;&amp; \
ln -sf &quot;libwasmer.so.$pkgver&quot; &quot;/nix/store/ckai343wfz8gd0xg47ajfxa7xbjzmfz6-wasmer-7.1.0/lib/libwasmer.so&quot;
install: cannot stat &#x27;target/release/libwasmer.so&#x27;: No such file or directory
make: *** [Makefile:962: install-capi-lib] Error 1</pre>
</details>
</details>

---
### `aarch64-darwin` (sandbox = relaxed)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>wasmer</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>wasmer</summary>
<pre>  cargo:warning=/nix/store/z7zz4h58ymzz8y28szyfsjzyhyby6nih-wasmer-v8-prebuilt-11.9.2/include/cppgc/macros.h:51:1: error: unknown type name &#x27;concept&#x27;
  cargo:warning=   51 | concept IsStackAllocatedType =
  cargo:warning=      | ^
  cargo:warning=/nix/store/z7zz4h58ymzz8y28szyfsjzyhyby6nih-wasmer-v8-prebuilt-11.9.2/include/cppgc/macros.h:52:5: error: use of undeclared identifier &#x27;requires&#x27;
  cargo:warning=   52 |     requires { typename T::IsStackAllocatedTypeMarker; };
  cargo:warning=      |     ^~~~~~~~
  cargo:warning=fatal error: too many errors emitted, stopping now [-ferror-limit=]
  cargo:warning=65 warnings and 20 errors generated.

  exit status: 1
  cargo:warning=ToolExecError: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL=&quot;C&quot; &quot;clang++&quot; &quot;-O3&quot; &quot;-ffunction-sections&quot; &quot;-fdata-sections&quot; &quot;-fPIC&quot; &quot;--target=arm64-apple-macosx&quot; &quot;-mmacosx-version-min=14.0&quot; &quot;-I&quot; &quot;/nix/store/z7zz4h58ymzz8y28szyfsjzyhyby6nih-wasmer-v8-prebuilt-11.9.2/include&quot; &quot;-I&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/include&quot; &quot;-I&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/v8/src&quot; &quot;-Wall&quot; &quot;-Wextra&quot; &quot;-DNAPI_EXTERN=&quot; &quot;-DV8_COMPRESS_POINTERS=1&quot; &quot;-o&quot; &quot;/nix/build/nix-9657-3393283968/source/target/release/build/wasmer-napi-bc2fb54f216afd52/out/a61a0efb84edfa4c-edge_v8_platform.o&quot; &quot;-c&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/v8/src/edge_v8_platform.cc&quot;

  --- stderr


  error occurred in cc-rs: command did not execute successfully (status code exit status: 1): env -u IPHONEOS_DEPLOYMENT_TARGET LC_ALL=&quot;C&quot; &quot;clang++&quot; &quot;-O3&quot; &quot;-ffunction-sections&quot; &quot;-fdata-sections&quot; &quot;-fPIC&quot; &quot;--target=arm64-apple-macosx&quot; &quot;-mmacosx-version-min=14.0&quot; &quot;-I&quot; &quot;/nix/store/z7zz4h58ymzz8y28szyfsjzyhyby6nih-wasmer-v8-prebuilt-11.9.2/include&quot; &quot;-I&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/include&quot; &quot;-I&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/v8/src&quot; &quot;-Wall&quot; &quot;-Wextra&quot; &quot;-DNAPI_EXTERN=&quot; &quot;-DV8_COMPRESS_POINTERS=1&quot; &quot;-o&quot; &quot;/nix/build/nix-9657-3393283968/source/target/release/build/wasmer-napi-bc2fb54f216afd52/out/a61a0efb84edfa4c-edge_v8_platform.o&quot; &quot;-c&quot; &quot;/nix/build/nix-9657-3393283968/source/lib/napi/v8/src/edge_v8_platform.cc&quot;


warning: build failed, waiting for other jobs to finish...
make: *** [Makefile:463: build-wasmer] Error 101</pre>
</details>
</details>

--- a/pkgs/by-name/wa/wasmer/package.nix
+++ b/pkgs/by-name/wa/wasmer/package.nix
@@ -2,30 +2,79 @@
   lib,
   stdenv,
   fetchFromGitHub,
+  fetchurl,
   rustPlatform,
   cargo,
   rustc,
-  llvmPackages_18,
+  nix-update,
+  curl,
+  writeShellApplication,
+  llvmPackages_21,
   libffi,
   libxml2,
   withLLVM ?
     stdenv.hostPlatform.isLinux || (stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isx86_64),
+  withV8 ? (stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isx86_64),
 }:
 
-stdenv.mkDerivation  (finalAttrs: {
+let
+  v8Version = "11.9.2";
+
+  # Prebuilt V8 from wasmerio's custom builds, only evaluated when withV8 = true.
+
+  # Per-platform hashes, auto-updated via the general updateScript
+  v8Hashes = {
+    "v8-linux-amd64.tar.xz" = "sha256-nTCVdBKtyVMb7lE+Db4RDsShKkLbG/0r980ejd+EAvo=";
+    "v8-linux-musl-amd64.tar.xz" = "sha256-XgRs3I46B2PG7Jrv5E+KSeuNfXLhgB7R66cAkA/Bvv8=";
+    "v8-darwin-arm64.tar.xz" = "sha256-xAG1PcAGw8a0A9k8d78/whTUXnqdfRZBz8yrg/+iz0M=";
+  };
+
+  v8Prebuilt =
+    let
+      assetName =
+        if stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isx86_64 && stdenv.hostPlatform.isMusl then
+          "v8-linux-musl-amd64.tar.xz"
+        else if stdenv.hostPlatform.isLinux && stdenv.hostPlatform.isx86_64 then
+          "v8-linux-amd64.tar.xz"
+        else if stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64 then
+          "v8-darwin-arm64.tar.xz"
+        else
+          throw "withV8 = true is not supported on ${stdenv.hostPlatform.system}";
+    in
+    stdenv.mkDerivation {
+      name = "wasmer-v8-prebuilt-${v8Version}";
+      src = fetchurl {
+        url = "https://github.com/wasmerio/v8-custom-builds/releases/download/${v8Version}/${assetName}";
+        hash = v8Hashes.${assetName};
+      };
+      sourceRoot = ".";
+      dontBuild = true;
+      installPhase = ''
+        cp -r . $out
+      '';
+
+      meta.sourceProvenance = [ lib.sourceTypes.binaryNativeCode ];
+    };
+in
+
+stdenv.mkDerivation (finalAttrs: {
   pname = "wasmer";
-  version = "6.0.1";
+  version = "7.1.0";
+
+  __structuredAttrs = true;
+  strictDeps = true;
 
   src = fetchFromGitHub {
     owner = "wasmerio";
     repo = "wasmer";
     tag = "v${finalAttrs.version}";
-    hash = "sha256-nw/4hcEDkAabcpatVBRozxvVLzYOKbj3ylrGeQtNzMQ=";
+    hash = "sha256-A1SkZY+iSR9xlu6R1p9uZYsGFPAOifuYTHtEXaEgves=";
+    fetchSubmodules = true;
   };
 
   cargoDeps = rustPlatform.fetchCargoVendor {
     inherit (finalAttrs) pname version src;
-    hash = "sha256-Nui8KxDk4+sqcmzeoZ6hGRb9Ux71+Nckz8seqq07cdE=";
+    hash = "sha256-wBEwGKjj9DdZESFlXS8T7B0Xdp7yMe08DYTGr4wnTVI=";
   };
 
   nativeBuildInputs = [
@@ -36,28 +85,71 @@ stdenv.mkDerivation  (finalAttrs: {
   ];
 
   buildInputs = lib.optionals withLLVM [
-    llvmPackages_18.llvm
+    llvmPackages_21.llvm
     libffi
     libxml2
   ];
 
+  # In 7.1.0 there is no nice flag to toggle napi/v8 on or off,
+  # so we manually delete the Makefile entry when we don't want v8
+  # TODO: v7.2.0 pre-release has a flag, when updating to 7.2.0
+  #  add "ENABLE_NAPI_V8=${if withV8 then "1" else "0"}" to makeFlags
+  postPatch =
+    lib.optionalString (!withV8) ''
+      substituteInPlace Makefile \
+        --replace-fail '	build_wasmer_extra_features += napi-v8' ""
+    ''
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
+      substituteInPlace Makefile \
+        --replace-fail 'install: install-wasmer install-capi-headers install-capi-lib install-pkgconfig install-misc' \
+                       'install: install-wasmer install-capi-headers install-misc'
+    '';
+
   makeFlags = [
     "WASMER_INSTALL_PREFIX=${placeholder "out"}"
     "DESTDIR=${placeholder "out"}"
-  ];
-
-  buildFlags = [
     "ENABLE_LLVM=${if withLLVM then "1" else "0"}"
   ];
 
-  env = lib.optionalAttrs withLLVM {
-    LLVM_SYS_180_PREFIX = llvmPackages_18.llvm.dev;
-  };
+  # Default all target includes headless C API which doesn't get installed
+  # by their Makefile and is quite niche to include, so we opt out
+  buildFlags = [
+    "build-wasmer"
+    "build-capi"
+  ];
+
+  env =
+    lib.optionalAttrs withLLVM {
+      LLVM_SYS_211_PREFIX = llvmPackages_21.llvm.dev;
+    }
+    // lib.optionalAttrs withV8 {
+      # build.rs skips the download when these are set; see resolve_explicit_v8 in lib/napi/build.rs
+      NAPI_V8_INCLUDE_DIR = "${v8Prebuilt}/include";
+      V8_LIB_DIR = "${v8Prebuilt}/lib";
+    };
+
+  postInstall = lib.optionalString stdenv.hostPlatform.isDarwin ''
+    install -Dm755 target/release/libwasmer.dylib $out/lib/libwasmer.dylib
+    if pc="$(WASMER_DIR="" target/release/wasmer config --pkg-config 2>/dev/null)"; then
+      mkdir -p "$out/lib/pkgconfig"
+      printf '%s\n' "$pc" > "$out/lib/pkgconfig/wasmer.pc"
+    fi
+  '';
+
+  passthru.updateScript = lib.getExe (writeShellApplication {
+    name = "update-wasmer";
+    runtimeInputs = [
+      nix-update
+      curl
+    ];
+    text = builtins.readFile ./update.sh;
+  });
 
   # Tests are failing due to `Cannot allocate memory` and other reasons
   doCheck = false;
 
   meta = {
+    sourceProvenance = with lib.sourceTypes; [ fromSource ] ++ lib.optional withV8 binaryNativeCode;
     description = "Universal WebAssembly Runtime";
     mainProgram = "wasmer";
     longDescription = ''

--- a/pkgs/by-name/wa/wasmer/update.sh
+++ b/pkgs/by-name/wa/wasmer/update.sh
@@ -1,0 +1,49 @@
+old_version=$(grep -oP '^  version = "\K[^"]+(?=";)' pkgs/by-name/wa/wasmer/package.nix)
+echo "Current wasmer version: $old_version"
+
+nix-update wasmer "$@"
+
+new_version=$(grep -oP '^  version = "\K[^"]+(?=";)' pkgs/by-name/wa/wasmer/package.nix)
+if [ "$old_version" = "$new_version" ]; then
+  echo "Already at $old_version, nothing to do"
+  exit 0
+fi
+echo "Updated wasmer $old_version -> $new_version"
+
+# v8Version is pinned in package.nix to match PREBUILT_V8_VERSION in lib/napi/build.rs.
+# lib/napi is a submodule, so resolve its pinned SHA first, then fetch build.rs from that repo.
+echo "Fetching PREBUILT_V8_VERSION from lib/napi/build.rs..."
+napi_sha=$(curl -fsSL "https://api.github.com/repos/wasmerio/wasmer/contents/lib/napi?ref=v${new_version}" \
+           | grep -oP '"sha":\s*"\K[^"]+' | head -1)
+new_v8=$(curl -fsSL "https://raw.githubusercontent.com/wasmerio/napi/${napi_sha}/build.rs" \
+         | grep -oP 'PREBUILT_V8_VERSION\s*:\s*&str\s*=\s*"\K[^"]+')
+cur_v8=$(grep -oP '^  v8Version = "\K[^"]+(?=";)' pkgs/by-name/wa/wasmer/package.nix)
+echo "V8: current=$cur_v8, required=$new_v8"
+
+if [ "$new_v8" = "$cur_v8" ]; then
+  echo "V8 version unchanged, done"
+  exit 0
+fi
+
+echo "V8 bumped $cur_v8 -> $new_v8, fetching hashes for all platforms..."
+sed -i "s|^  v8Version = \"[^\"]*\";$|  v8Version = \"$new_v8\";|" \
+  pkgs/by-name/wa/wasmer/package.nix
+
+base="https://github.com/wasmerio/v8-custom-builds/releases/download/$new_v8"
+declare -A assets=(
+  ["v8-linux-amd64.tar.xz"]="$base/v8-linux-amd64.tar.xz"
+  ["v8-linux-musl-amd64.tar.xz"]="$base/v8-linux-musl-amd64.tar.xz"
+  ["v8-darwin-arm64.tar.xz"]="$base/v8-darwin-arm64.tar.xz"
+)
+
+for asset in "${!assets[@]}"; do
+  url="${assets[$asset]}"
+  echo "  Fetching hash for $asset..."
+  hash=$(nix-prefetch-url --type sha256 "$url" 2>/dev/null \
+         | xargs nix hash convert --hash-algo sha256 --to sri)
+  echo "  $asset -> $hash"
+  sed -i "s|\"$asset\" = \"[^\"]*\"|\"$asset\" = \"$hash\"|" \
+    pkgs/by-name/wa/wasmer/package.nix
+done
+
+echo "Done"

